backend/queues/makeOfferProcessor.js

```diff
@@ -34,48 +57,126 @@ async function processor(job) {
-    const network = await getNetwork(shop.networkId)
+    const network = await _getNetwork(shop.networkId)
     const networkConfig = encConf.getConfig(network.config)
     const shopConfig = encConf.getConfig(shop.config)
 
     queueLog(10, 'Creating offer')
     const lid = ListingID.fromFQLID(shop.listingId)
-    const offer = createOfferJson(lid, encryptedData, paymentCode)
-    const ires = await postOfferIPFS(network, offer)
+    const offer = _createOfferJson(lid, encryptedData, paymentCode)
+    const ipfsHash = await _postOfferIPFS(network, offer)
 
     queueLog(20, 'Submitting Offer')
-    const web3 = new Web3(network.provider)
 
     // Use the Shop PK if there is one, otherwise fall back to Network PK.
     if (!shopConfig.web3Pk) {
       if (!networkConfig.web3Pk) {
-        throw new Error('PK missing in both shop and network configs')
+        throw new Error(`PK missing in both shop ${shopId} and network configs`)
       }
       log.info(
         `No PK configured for shop ${shopId}. Falling back to network PK`
       )
     }
-    const backendPk = shopConfig.web3Pk || networkConfig.web3Pk
-    const account = web3.eth.accounts.wallet.add(backendPk)
-    const walletAddress = account.address
-    queueLog(22, `Using walletAddress ${walletAddress}`)
-    queueLog(25, 'Sending to marketplace')
-    const tx = await offerToMarketplace(
-      web3,
-      lid,
-      network,
-      walletAddress,
-      offer,
-      ires
+    const pk = shopConfig.web3Pk || networkConfig.web3Pk
+    const provider = new ethers.providers.JsonRpcProvider(network.provider)
+    const wallet = new ethers.Wallet(pk, provider)
+    const walletAddress = wallet.address
+    const marketplace = new ethers.Contract(
+      network.marketplaceContract,
+      marketplaceAbi,
+      wallet
     )
-    queueLog(50, JSON.stringify(tx))
+    queueLog(22, `Using walletAddress ${walletAddress}`)
+
+    let tx, transaction
 
-    // TODO: Code to prevent duplicate txs
-    // TODO Record tx and wait for TX to go through the blockchain
+    // Check if there is a pending transaction for the wallet.
+    // It's possible the processing got interrupted while waiting for the tx to get mined.
+    // For example due to a server maintenance or a crash.
+    transaction = await Transaction.findOne({
+      where: {
+        networkId: network.networkId,
+        wallet: walletAddress,
+        status: TransactionStatuses.Pending
+      }
+    })
+    if (transaction) {
+      log.info(
+        `Found pending transaction ${transaction.id} job ${job.id} hash ${transaction.hash} for wallet ${walletAddress}`
+      )
+
+      // If it is not the transaction from our job. Do not try to recover it.
+      // Let it get recovered by the job that created it.
+      // Fail our job for now, it will get retried.
+      if (transaction.jobId === job.id) {
+        throw new Error(
+          `Pending transaction does not belongs to job ${job.id}. Bailing.`
```

Comment:
What happens for busy shops that say have 5 pending orders?  Does this job just fail repeatedly until the retry limit is reached?

Reply:
Good catch. I meant to use the !== operator. I will also add a unit test.
The idea is that the pending transaction will only get processed by the retry for the jobId that created it.
Other transactions will throw exception and get retried and should eventually succeed as part of a retry once the pending transaction has been mined.

Reply:
A thrown exception will count as a retry for the job though, wouldn't it?  This could cause it to fail say 3 times before the transaction is actually returned from that query.